src/ucp/rma/basic_rma.c

```diff
@@ -75,6 +77,156 @@ ucs_status_t ucp_put(ucp_ep_h ep, const void *buffer, size_t length,
     return status;
 }
 
+static ucs_status_t ucp_progress_put_nbi(uct_pending_req_t *self)
+{
+
+    ucs_status_t status;
+    ssize_t packed_len;
+
+    ucp_request_t *req = ucs_container_of(self, ucp_request_t, send.uct);
+
+    ucp_ep_t *ep = req->send.ep;
+    uct_rkey_t uct_rkey = UCP_RKEY_LOOKUP(ep, req->send.rma.rkey);
+
+    for (;;) {
+        if (req->send.length <= ep->worker->context->config.bcopy_thresh) {
+            /* Should be replaced with bcopy */
+            packed_len = ucs_min(req->send.length, ep->config.max_short_put);
+            status = uct_ep_put_short(ep->uct_ep,
+                                      req->send.buffer,
+                                      packed_len,
+                                      req->send.rma.remote_addr,
+                                      uct_rkey);
+        } else {
+            /* We don't do it right now, but in future we have to add
+             * an option to use zcopy
+             */
+            size_t frag_length;
+            ucp_memcpy_pack_context_t pack_ctx;
+            pack_ctx.src    = req->send.buffer;
+            pack_ctx.length = frag_length =
+                ucs_min(req->send.length, ep->config.max_bcopy_put);
+            packed_len = uct_ep_put_bcopy(ep->uct_ep,
+                                          ucp_memcpy_pack,
+                                          &pack_ctx,
+                                          req->send.rma.remote_addr,
+                                          uct_rkey);
+            status = (packed_len > 0) ? UCS_OK : (ucs_status_t)packed_len;
+        }
+
+        if (ucs_likely(status == UCS_OK || status == UCS_INPROGRESS)) {
+            req->send.length -= packed_len;
+            if (req->send.length == 0) {
+                ucp_request_rma_complete(req);
+                break;
+            }
+
+            req->send.buffer += packed_len;
+            req->send.rma.remote_addr += packed_len;
+        } else {
+            break;
+        }
+    }
+
+    return status;
```

Comment:
it's not enough to return from pending function. You also have to make sure ucp_worker_flush() does not return before this happens (e.g pending requests are flushed). Please note that *UCT flush does not flush pending* to avoid possible deadlocks with 2-sided protocols. One way to do it is to add counter for outstanding RMAs.